File: technical-tests/backend/milan-jovanovic/source-code-async-locking/after/Currency.Converter/Currency.Converter/CurrencyConversion.cs
using System.Collections.Concurrent;

namespace Currency.Converter;

public static class CurrencyConversion
{
    public static async Task<IResult> Handle(
        string currencyCode,
        decimal amount,
        CurrencyApiClient currencyClient)
    {
        // Validate currency code format (3 uppercase letters)
        if (string.IsNullOrWhiteSpace(currencyCode) ||
            currencyCode.Length != 3 ||
            !currencyCode.All(char.IsLetter))
        {
            return Results.BadRequest(
                new { error = "Currency code must be a 3-letter uppercase code (e.g., EUR, GBP)" });
        }

        // Validate amount (must be positive)
        if (amount < 0)
        {
            return Results.BadRequest(new { error = "Amount must be a positive number" });
        }

        var rate = await GetExchangeRate(currencyCode, currencyClient);

        if (rate == null)
        {
            return Results.NotFound(
                new { error = $"Exchange rate for {currencyCode} not found or API error occurred" });
        }

        var convertedAmount = amount * rate.Value;

        return Results.Ok(new ExchangeRateResponse(
            Currency: currencyCode,
            BaseCurrency: "USD",
            Rate: rate.Value,
            Amount: amount,
            ConvertedAmount: convertedAmount
        ));
    }

    private record CacheEntry(decimal Rate, DateTime CreatedAtUtc);
    private static readonly ConcurrentDictionary<string, CacheEntry> Cache = new();
    private static readonly TimeSpan CacheDuration = TimeSpan.FromMinutes(5);
    //private static readonly SemaphoreSlim Semaphore = new(1, 1);
    private static readonly ConcurrentDictionary<string, SemaphoreSlim> Locks = new();

    private static async Task<decimal?> GetExchangeRate(
        string currencyCode,
        CurrencyApiClient currencyClient)
    {
        if (Cache.TryGetValue(currencyCode, out var entry) && IsFresh(entry))
        {
            return entry.Rate;
        }

        var semaphore = Locks.GetOrAdd(currencyCode, _ => new SemaphoreSlim(1, 1));

        var acquired = await semaphore.WaitAsync(TimeSpan.FromSeconds(5));
        if (!acquired)
        {
            throw new TimeoutException("Could not acquired exchange rates. Try again later");
        }

        try
        {
            if (Cache.TryGetValue(currencyCode, out entry) && IsFresh(entry))
            {
                return entry.Rate;
            }

            var currentRate = await currencyClient.GetExchangeRateAsync(currencyCode);

            if (currentRate != null)
            {
                Cache.AddOrUpdate(
                    currencyCode,
                    _ => new CacheEntry(currentRate.Value, DateTime.UtcNow),
                    (_, _) => new CacheEntry(currentRate.Value, DateTime.UtcNow));
            }

            return currentRate;
        }
        finally
        {
            semaphore.Release();
        }
    }

    private static bool IsFresh(CacheEntry entry)
    {
        return DateTime.UtcNow - entry.CreatedAtUtc < CacheDuration;
    }
}

File: technical-tests/backend/milan-jovanovic/source-code-async-locking/typescript-after/currency-conversion.ts
import { Semaphore } from "../../../../../semaphore-mutex";
import { CurrencyApiClient } from "./currency-api-client";
import { ExchangeRateResponse } from "./exchange-rate-response";

export class CacheEntry {
  rate: number;
  createdAt: Date;
}

export class CurrencyConversion {
  private readonly cache: Map<string, CacheEntry> = new Map();
  private readonly cacheDuration: number = 5 * 60 * 1000;
  private readonly locks: Map<string, Semaphore> = new Map();
  async handle(
    currencyCode: string,
    amount: number,
    currencyClient: CurrencyApiClient,
  ): Promise<ExchangeRateResponse> {
    if (
      !currencyCode ||
      currencyCode.length !== 3 ||
      !currencyCode.toUpperCase().match(/^[A-Z]{3}$/)
    ) {
      throw new Error(
        "Currency code must be a 3-letter uppercase code (e.g., EUR, GBP)",
      );
    }
    if (amount < 0) {
      throw new Error("Amount must be a positive number");
    }

    const rate = await this.getExchangeRate(currencyCode, currencyClient);
    if (rate === undefined || rate === null) {
      throw new Error(
        `Exchange rate for ${currencyCode} not found or API error occurred`,
      );
    }

    const convertedAmount = amount * rate;
    return {
      currency: currencyCode,
      baseCurrency: "USD",
      rate: rate,
      amount: amount,
      convertedAmount: convertedAmount,
    };
  }

  private isFresh(entry: CacheEntry): boolean {
    return Date.now() - entry.createdAt.getTime() < this.cacheDuration;
  }

  private async getExchangeRate(
    currencyCode: string,
    currencyClient: CurrencyApiClient,
  ): Promise<number> {
    const entry = this.cache.get(currencyCode);
    if (entry && this.isFresh(entry)) {
      return entry.rate;
    }

    const semaphore = this.locks.get(currencyCode) || new Semaphore(1);
    this.locks.set(currencyCode, semaphore);

    const acquired = await Promise.race([
      semaphore.acquire(),
      new Promise((_, reject) =>
        setTimeout(() => reject(new Error("Timeout")), 5000),
      ),
    ]);

    if (!acquired) {
      throw new Error("Timeout");
    }

    try {
      const entry = this.cache.get(currencyCode);
      if (entry && this.isFresh(entry)) {
        return entry.rate;
      }

      const rate = await currencyClient.getExchangeRate(currencyCode);

      if (rate) {
        this.cache.set(currencyCode, { rate: rate, createdAt: new Date() });
        return rate;
      }

      if (!rate) throw new Error("Rate doesnt exist");

      return rate;
    } catch (error) {
      throw new Error(
        `Error fetching exchange rate for ${currencyCode}: ${error.message}`,
      );
    } finally {
      semaphore.release();
    }
  }
}
